proto/storage/auth_provider.proto

```diff
@@ -33,6 +33,9 @@ message AuthProvider {
         string attribute_key   = 1;
         string attribute_value = 2;
     }
-    // EXPERIMENTAL.
     repeated RequiredAttribute required_attributes = 11;
+
+    // Modification via UI/API and deletion via UI are forbidden for this auth provider.
+    // Deletion via API is allowed only when `force` flag is set to true.
+    bool unmodifiable = 12;
```

Comment:
I share the sentiment that there should be a more general way of highlighting where an object comes from/is stored. Given the discussion around possibly injecting checks at the SQL level, I am not sure if this wouldn't be defeating the purpose of doing something quick for service preview.

After all, is there really much utility in having objects that are unmodifiable stored in the SQL database? The way I was thinking about it, we would have a general datastore interface that provides an aggregated view onto a read-only in-memory store populated (and updated) from mounted declarative config layered on top of a SQL-based store.

This can get fairly complicated - we probably need code generation for both the in-memory store and the aggregation layer, some less-than-obvious for deciding to which underlying store writes/creates are directed, and a method to communicate to the UI where an object comes from/if it is modifiable (even though the last two could possibly be combined with a `source` or `origin` field).

If we store objects in the database, we need to support reconciliation efficiently. I.e., if the configmap gets deleted, we need to know which entities need to be deleted from the DB. If central comes back up, we need to have a way to maintain ID persistence (do we expect the user to specify a UUID in the configmap entry?). We need to have all kinds of rules how an object with a certain `origin` can be created, modified, deleted, but in the end, we take everything from a "dumb" SQL database without business logic, where it isn't clear how to deal with inconsistencies.


All in all, there are a lot of things to think about here. The use case of having unmodifiable objects that however can be created and deleted through the API might well be an artifact - and it _is_ weird in all kinds of ways, such as introducing an API parameter we seek to soon remove again, even though we generally want our APIs to be stable; or the fact that "unmodifiable" actually means "well, not really, but you have to say 'pretty please'". I do not see it as a stepping stone to supporting declarative configuration, to be honest, and that is not because of the implementation in this PR, but because it conceptually doesn't make a whole lot of sense.

There might be some way of evolving it to supporting the declarative configuration use case, but my current assessment would be that we will deprecate/delete everything related to this mechanism once we have support for declarative configuration. Looking to generalize here might not only slow down the implementation of the quick (albeit conceptually flawed) solution that was agreed upon, but also complicate the design of general support for declarative config objects, without carrying over any value.

Reply:
Thank you, Malte, for your quick and elaborate response, I appreciate the details.

I spoke to another advocate of declarative configuration, @connorgorman, and indeed, looks like we need to do some thinking on how we marry DB objects with declaratively created ones. With this in mind, I agree that designing an extensible solution will likely become poorly invested time.

With this out of the way, @ivan-degtiarenko let's do the following:
* introduce `storage/proto/traits.proto` with the following (we'll need the same for groups): 
    ```
    // EXPERIMENTAL.
    // 
    // MutabilityMode specifies whether and how an object can be modified. Default
    // is ALLOW and means there are no modification restrictions; this is equivalent
    // to the absence of MutabilityMode specification. ALLOW_FORCED forbids all
    // modifying operations except object removal with force bit on.
    //
    // Be careful when changing the state of this field. For example, modifying an
    // object from ALLOW to ALLOW_FORCED is allowed but will prohibit any further
    // changes to it, including modifying it back to ALLOW.
    enum MutabilityMode {
        ALLOW        = 0;
        ALLOW_FORCED = 1;
    }
    
    // EXPERIMENTAL.
    message Traits {
        MutabilityMode mutability = 1;
    }
    
    // EXPERIMENTAL.
    //
    // Used in combination with MutabilityMode.ALLOW_FORCED to enable forced removal.
    message DeleteByIDWithForce {
        string id  = 1;
        bool force = 2;
    }
    ```
* Mark `traits` field in auth provider as experimental as well, explaining how it works 
* Move the checks to the datastore level
* Cover the functionality with tests, including cases like not allowing to change the mutability mode from `ALLOW_FORCE` to `ALLOW`, forced and non-forced removal, updates to `ALLOW` and `ALLOW_FORCE` objects, etc

Reply:
Thank you for your collaborative effort on this, folks.

@rukletsov what is the use case to support `ALLOW_FORCED` mutability mode in groups?

AFAIR, the original idea was for `sso.redhat.com` auth provider groups to be open to modification - otherwise, customer would need to integrate with separate IdP to change default role etc.
SREs `sso.redhat.com` auth provider(if one will exist) will be hidden and also will not benefit from `ALLOW_FORCED` groups.

Reply:
@ivan-degtiarenko if don't do that, users can remove both admin groups and lock themselves out. The idea is the same: all resource objects created during the bootstrap shall be immutable (aka unmodifiable aka ALLOW_FORCE). Any group for this auth provider added on top will not be immutable.